contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,506 @@
+                    token_in,
+                    token_out,
+                    amount_in: amount_token_in,
+                    amount_out: amount_token_out,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Liquidity deposit
+        ///
+        /// Can only be performed by an account with a LiquidityProvider role
+        /// Caller needs to give at least the passed amount of allowance to the contract to spend the deposited tokens on his behalf
+        /// prior to executing this tx
+        #[ink(message)]
+        pub fn deposit(&mut self, deposits: Vec<(AccountId, Balance)>) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check role, only designated account can add liquidity
+            self.check_role(caller, Role::LiquidityProvider(this))?;
+
+            deposits
+                .into_iter()
+                .try_for_each(|(token_in, amount)| -> Result<(), DexError> {
+                    // transfer token_in from the caller to the contract
+                    // will revert if the contract does not have enough allowance from the caller
+                    // in which case the whole tx is reverted
+                    self.transfer_from_tx(token_in, caller, this, amount)??;
+                    Ok(())
+                })?;
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn withdrawal(
+            &mut self,
+            withdrawals: Vec<(AccountId, Balance)>,
+        ) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check role, only designated account can remove liquidity
+            self.check_role(caller, Role::LiquidityProvider(this))?;
+
+            withdrawals.into_iter().try_for_each(
+                |(token_out, amount)| -> Result<(), DexError> {
+                    // transfer token_out from the contract to the caller
+                    self.transfer_tx(token_out, caller, amount)??;
+                    Ok(())
+                },
+            )?;
+
+            Ok(())
+        }
+
+        /// Alters the swap_fee parameter
+        ///
+        /// Can only be called by the contract's Admin.
+        #[ink(message)]
+        pub fn set_swap_fee_percentage(
+            &mut self,
+            swap_fee_percentage: u128,
+        ) -> Result<(), DexError> {
+            if swap_fee_percentage.gt(&100) {
+                return Err(DexError::WrongParameterValue);
+            }
+
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            self.check_role(caller, Role::Admin(this))?;
+
+            // emit event
+            Self::emit_event(
+                self.env(),
+                Event::SwapFeeSet(SwapFeeSet {
+                    caller,
+                    swap_fee_percentage,
+                }),
+            );
```

Comment:
nit: I would think that event is emitted after the action has happened. In practice it wouldn't matter probably but it's a matter of personal style more than anything.